File: Data Structures & Algorithms/climbing-stairs/submission-3.py
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = {}

        def dfs(i):
            # Base conditions
            if i == 0 or i == 1 or i == 2:
                dp[i] = i
                return i

            if i in dp:
                return dp[i]

            dp[i] = dfs(i - 1) + dfs(i - 2)
            return dp[i]

        return dfs(n)

File: Data Structures & Algorithms/balanced-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool dfs(TreeNode *root, int &h) {
        int lh = 0;
        int rh = 0;

        if (root == nullptr) {
            return true;
        }

        bool l_balanced = dfs(root->left, lh);
        bool r_balanced = dfs(root->right, rh);
        h = 1 + max(lh, rh);

        if (l_balanced && r_balanced && (abs(lh - rh) <= 1)) {
            return true;
        } else {
            return false;
        }

    }

    bool isBalanced(TreeNode* root) {
        int h = 0;
        return dfs(root, h);
    }
};
